include/hip/hcc_detail/program_state.hpp

```diff
@@ -558,92 +561,156 @@ const std::unordered_map<
-    std::vector<std::pair<std::size_t, std::size_t>>& size_align) {
-    if (f == l) return f;
-    if (!size_align.empty()) return l;
-
-    do {
-        static constexpr size_t size_sz{5};
-        f = metadata.find("Size:", f) + size_sz;
-
-        if (l <= f) return f;
-
-        auto size = std::strtoul(&metadata[f], nullptr, 10);
-
-        static constexpr size_t align_sz{6};
-        f = metadata.find("Align:", f) + align_sz;
-
-        char* l{};
-        auto align = std::strtoul(&metadata[f], &l, 10);
-
-        f += (l - &metadata[f]) + 1;
+void checkError(
+    amd_comgr_status_t status,
+    char const *str) {
+    if (status != AMD_COMGR_STATUS_SUCCESS) {
+        const char *status_str;
+        status = amd_comgr_status_string(status, &status_str);
+        std::cerr << "FAILED: " << str << "\n  REASON: " <<  status_str << std::endl;
+        hip_throw(std::runtime_error{"Metadata parsing failed."});
+    }
+}
 
-        size_align.emplace_back(size, align);
-    } while (true);
+inline
+std::string lookup_keyword_value(
+    amd_comgr_metadata_node_t& in_node,
+    std::string keyword) {
+    amd_comgr_status_t status;
+    size_t value_size;
+    amd_comgr_metadata_node_t value_node;
+
+    status = amd_comgr_metadata_lookup(in_node, keyword.c_str(), &value_node);
+    checkError(status, "amd_comgr_metadata_lookup");
+    status = amd_comgr_get_metadata_string(value_node, &value_size, NULL);
+    checkError(status, "amd_comgr_get_metadata_string");
+    // Since value_size returns size with null terminator, we don't include for C++ string size
+    value_size--;
+    std::string value(value_size, '\0');
+    status = amd_comgr_get_metadata_string(value_node, &value_size, &value[0]);
+    checkError(status, "amd_comgr_get_metadata_string");
+
+    status = amd_comgr_destroy_metadata(value_node);
+    checkError(status, "amd_comgr_destroy_metadata");
+    return value;
 }
 
 inline
-void read_kernarg_metadata(
-    ELFIO::elfio& reader,
+void process_kernarg_metadata(
+    amd_comgr_metadata_node_t blob_meta,
     std::unordered_map<
         std::string,
         std::vector<std::pair<std::size_t, std::size_t>>>& kernargs) {
-    // TODO: this is inefficient.
-    auto it = find_section_if(reader, [](const ELFIO::section* x) {
-        return x->get_type() == SHT_NOTE;
-    });
+    amd_comgr_status_t status;
+    amd_comgr_metadata_kind_t mkindLookup;
+    amd_comgr_metadata_node_t kernelList, kernelMap;
+    amd_comgr_metadata_node_t kernArgList, kernArgMap;
+    std::string kernel_name;
+    size_t num_kernels = 0;
+    size_t num_kern_args = 0;
+
+    // Kernels is a list of MAPS!!
+    status = amd_comgr_metadata_lookup(blob_meta, "Kernels", &kernelList);
+    if(status != AMD_COMGR_STATUS_SUCCESS)
+        return;
+
+    status = amd_comgr_get_metadata_kind(kernelList, &mkindLookup);
+    if (mkindLookup != AMD_COMGR_METADATA_KIND_LIST) {
+        hip_throw(std::runtime_error{"Lookup of Kernels didn't return a list\n"});
```

Comment:
Is there no way to recover from this throw? If there is, `kernelList` will leak here. Similar question for the other throws below.

Reply:
I could modify hip_throw to clean up all nodes that are created. When I pass a declared (but not used) node into amd_comgr_destroy_metadata, it will seg fault. Would it be good to keep a vector of created nodes, and destroy each of them on hip_throw?

Reply:
Tracking each initialized object independently (either via a vector as you mention, or by wrapping them in a simple class containing the handle and a boolean) is the only way to mange this without the potential to leak. The Comgr API does not allow for inspecting/modifying the opaque handle itself, so it really has to be separate.

Reply:
I've changed the implementation to use simple class as you've suggested, and the destructor will destroy on hip_throw and leaving of scope.